File: generated-code/cpp/codecraft/FileReadWrite/model/debug_interface/ColoredVertex.cpp
#include "ColoredVertex.hpp"

namespace model::debug_interface {

ColoredVertex::ColoredVertex(std::optional<Vec2Float> worldPos, Vec2Float screenOffset, Color color) : worldPos(worldPos), screenOffset(screenOffset), color(color) { }

// Read ColoredVertex from input stream
ColoredVertex ColoredVertex::readFrom(InputStream& stream) {
    std::optional<Vec2Float> worldPos = std::optional<Vec2Float>();
    if (stream.readBool()) {
        worldPos = Vec2Float::readFrom(stream);
    }
    Vec2Float screenOffset = Vec2Float::readFrom(stream);
    Color color = Color::readFrom(stream);
    return ColoredVertex(worldPos, screenOffset, color);
}

// Write ColoredVertex to output stream
void ColoredVertex::writeTo(OutputStream& stream) const {
    if (worldPos) {
        stream.write(true);
        const Vec2Float& worldPosValue = *worldPos;
        worldPosValue.writeTo(stream);
    } else {
        stream.write(false);
    }
    screenOffset.writeTo(stream);
    color.writeTo(stream);
}

// Get string representation of ColoredVertex
std::string ColoredVertex::toString() const {
    std::stringstream ss;
    ss << "ColoredVertex { ";
    ss << "worldPos: ";
    if (worldPos) {
        const Vec2Float& worldPosValue = *worldPos;
        ss << worldPosValue.toString();
    } else {
        ss << "none";
    }
    ss << ", ";
    ss << "screenOffset: ";
    ss << screenOffset.toString();
    ss << ", ";
    ss << "color: ";
    ss << color.toString();
    ss << " }";
    return ss.str();
}

}

File: generated-code/rust/codecraft/FileReadWrite/src/model/debug_interface/colored_vertex.rs
use super::*;

/// Vertex for debug rendering
#[derive(Clone, Debug)]
pub struct ColoredVertex {
    /// Position in world coordinates (if none, screen position (0, 0) is used)
    pub world_pos: Option<Vec2F32>,
    /// Additional offset in screen coordinates
    pub screen_offset: Vec2F32,
    /// Color to use
    pub color: Color,
}

impl trans::Trans for ColoredVertex {
    fn write_to(&self, writer: &mut dyn std::io::Write) -> std::io::Result<()> {
        self.world_pos.write_to(writer)?;
        self.screen_offset.write_to(writer)?;
        self.color.write_to(writer)?;
        Ok(())
    }
    fn read_from(reader: &mut dyn std::io::Read) -> std::io::Result<Self> {
        let world_pos: Option<Vec2F32> = trans::Trans::read_from(reader)?;
        let screen_offset: Vec2F32 = trans::Trans::read_from(reader)?;
        let color: Color = trans::Trans::read_from(reader)?;
        Ok(Self {
            world_pos,
            screen_offset,
            color,
        })
    }
}